x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentNodeService.java

```diff
@@ -188,39 +197,68 @@ void loadQueuedModels() {
             );
             if (unassignedIndices.size() > 0) {
                 logger.trace("not loading models as indices {} primary shards are unassigned", unassignedIndices);
+                onFinish.run();
                 return;
             }
         }
-        logger.trace("attempting to load all currently queued models");
-        // NOTE: As soon as this method exits, the timer for the scheduler starts ticking
-        Deque<TrainedModelDeploymentTask> loadingToRetry = new ArrayDeque<>();
-        while ((loadingTask = loadingModels.poll()) != null) {
-            final String deploymentId = loadingTask.getDeploymentId();
-            if (loadingTask.isStopped()) {
-                if (logger.isTraceEnabled()) {
-                    String reason = loadingTask.stoppedReason().orElse("_unknown_");
-                    logger.trace("[{}] attempted to load stopped task with reason [{}]", deploymentId, reason);
+
+        // deploy the models synchronously by flattening loadingModels into a chain of listeners
+        // if we need to retry, wait to add it back to loadingModels until after we finish this batch of deployments
+        // if someone calls stop halfway through, abandon this entire chain
+        var loadingToRetry = new ConcurrentLinkedDeque<TrainedModelDeploymentTask>();
+        var deploymentChain = SubscribableListener.<Void>newSucceeded(null);
+        while (loadingModels.isEmpty() == false) {
+            var loadingTask = loadingModels.poll();
+            deploymentChain = deploymentChain.andThen((l, r) -> {
+                if (stopped) {
+                    // don't bother calling the listener, the lifecycle will not resume the instance of this class
+                    return;
```

Comment:
I'd sorta recommend just completing the chain of listeners in this case unless there's a strong argument (e.g. performance) not to do so, and then putting a `stopped` check in the final listener too. Leaking listeners is a source of super-painful bugs, even at shutdown.